server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -284,11 +285,21 @@ private void ensureRefCount() {
     /*** Implementation of {@link RecoveryTargetHandler } */
 
     @Override
-    public void prepareForTranslogOperations(boolean fileBasedRecovery, int totalTranslogOps, ActionListener<Void> listener) {
+    public void prepareForTranslogOperations(boolean fileBasedRecovery, int totalTranslogOps, long recoverUpToSeqNo,
+                                             ActionListener<Optional<Store.MetadataSnapshot>> listener) {
         ActionListener.completeWith(listener, () -> {
-            state().getTranslog().totalOperations(totalTranslogOps);
-            indexShard().openEngineAndSkipTranslogRecovery();
-            return null;
+            if (Assertions.ENABLED && recoverUpToSeqNo != SequenceNumbers.UNASSIGNED_SEQ_NO) {
+                final long gcp = readGlobalCheckpointFromTranslog();
+                assert recoverUpToSeqNo == gcp : recoverUpToSeqNo + " != " + gcp;
+            }
+            indexShard().openEngineAndRecoverFromTranslog(recoverUpToSeqNo);
+            if (indexShard.getLocalCheckpoint() < recoverUpToSeqNo) {
+                // This can happen if the previous recovery was aborted after copying segments, then the target does not have all translog.
+                indexShard.closeEngine();
+                return Optional.of(indexShard.snapshotStoreMetadata());
+            }
+            state().getTranslog().totalOperations(state().getTranslog().totalOperations() + totalTranslogOps);
+            return Optional.empty();
```

Comment:
Should we commit here? It'd save us from having to replay these operations again if something goes wrong from here on.